url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/ReflectionTemplateBuilder.java

```diff
@@ -0,0 +1,204 @@
+            Object o = template.read(unpacker, f, required);
+            if (o != f) {
+                entry.set(to, o);
+            }
+            return o;
+        }
+    }
+
+    protected static class ReflectionClassTemplate<T> extends AbstractTemplate<T> {
+        protected Class<T> targetClass;
+
+        protected ReflectionFieldTemplate[] templates;
+
+        protected ReflectionClassTemplate(Class<T> targetClass, ReflectionFieldTemplate[] templates) {
+            this.targetClass = targetClass;
+            this.templates = templates;
+        }
+
+        
+        public void write(Packer packer, T target, boolean required)
+                throws IOException {
+            if (target == null) {
+                if (required) {
+                    throw new MessageTypeException("attempted to write null");
+                }
+                packer.writeNil();
+                return;
+            }
+            try {
+                packer.writeArrayBegin(templates.length);
+                for (ReflectionFieldTemplate tmpl : templates) {
+                    if (!tmpl.entry.isAvailable()) {
+                        packer.writeNil();
+                        continue;
+                    }
+                    Object obj = tmpl.entry.get(target);
+                    if (obj == null) {
+                        if (tmpl.entry.isNotNullable()) {
+                            throw new MessageTypeException(tmpl.entry.getName()
+                                    + " cannot be null by @NotNullable");
+                        }
+                        packer.writeNil();
+                    } else {
+                        tmpl.write(packer, obj, true);
+                    }
+                }
+                packer.writeArrayEnd();
+            } catch (IOException e) {
+                throw e;
+            } catch (Exception e) {
+                throw new MessageTypeException(e);
+            }
+        }
+
+        
+        public T read(Unpacker unpacker, T to, boolean required)
+                throws IOException {
+            if (!required && unpacker.trySkipNil()) {
+                return null;
+            }
+            try {
+                if (to == null) {
+                    to = targetClass.newInstance();
+                }
+
+                unpacker.readArrayBegin();
+                for (int i = 0; i < templates.length; i++) {
+                    ReflectionFieldTemplate tmpl = templates[i];
+                    if (!tmpl.entry.isAvailable()) {
+                        unpacker.skip();
+                    } else if (tmpl.entry.isOptional() && unpacker.trySkipNil()) {
+                        // if Optional + nil, than keep default value
+                    } else {
+                        tmpl.read(unpacker, to, false);
+                    }
+                }
+
+                unpacker.readArrayEnd();
+                return to;
+            } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)